src/main/groovy/edu/oregonstate/mist/personsapi/PersonsResource.groovy

```diff
@@ -746,6 +749,122 @@ class PersonsResource extends Resource {
+            [false, "addressLine3", 75, null],
+            [false, "addressLine4", 75, null],
+            [true, "city", 50, null],
+            [false, "countyCode", 5,
+             { String countyCode -> bannerPersonsReadDAO.isValidCountyCode(countyCode) }
+            ],
+            [false, "stateCode", 3,
+             { String stateCode -> bannerPersonsReadDAO.isValidStateCode(stateCode) }
+            ],
+            [false, "postalCode", 30, null],
+            [false, "nationCode", 5,
+             { String nationCode -> bannerPersonsReadDAO.isValidNationCode(nationCode) }
+            ]
+        ].each {
+            Boolean isRequired = it.get(0)
+            String fieldName = it.get(1)
+            String fieldValue = address[fieldName]
+            Integer length = it.get(2)
+            Closure validateFunction = it.get(3)
+
+            // Check if requried field is missing
+            if (isRequired && !fieldValue) {
+                addBadRequest("Required field $fieldName is missing.")
+            }
+
+            // Check if input field is over the buffer size
+            if (fieldValue?.length() > length) {
+                addBadRequest("$fieldName can't be more than $length characters.")
+            }
+
+            // Check if input field is valid
+            if (fieldValue && validateFunction && !validateFunction(fieldValue)) {
+                addBadRequest("$fieldName is not valid.")
+            }
+        }
+
+        errors
+    }
+
+    @Timed
+    @POST
+    @Consumes (MediaType.APPLICATION_JSON)
+    @Path('{osuID: [0-9]+}/addresses')
+    Response createAddress(@PathParam('osuID') String osuID,
+                           @Valid ResultObject resultObject) {
+        String pidm = bannerPersonsReadDAO.personExist(osuID)
+        if (!pidm) {
+            return notFound().build()
+        }
+        AddressObject address
+        try {
+            address = AddressObject.fromResultObject(resultObject)
+            if (!address) {
+                return badRequest("No address object provided.").build()
+            }
+
+            List<Error> errors = newAddressErrors(address)
+            if (errors) {
+                return errorArrayResponse(errors)
+            }
+        } catch (PersonObjectException e) {
+            return badRequest(
+                "Unable to parse address object or required fields are missing. " +
+                "Please make sure all required fields are included and in the corret format."
+            ).build()
+        }
+
+        String addressType = resultObject.data['attributes']['addressType']
+        AddressRecordObject addressRecord = bannerPersonsReadDAO.hasSameAddressType(
+            pidm, addressType
+        )
+
+        try {
+            if (addressRecord?.rowID) {
+                logger.info("Address with the same type exist. Deactivate the current one.")
+                bannerPersonsWriteDAO.deactivateAddress(pidm, addressRecord)
```

Comment:
We could deactivate an address and then fail to create a new address. We should reactive the address on a failure.

Reply:
@fernmig that's a good point, I thought about it too. But I think the issue here is to deactivate and reactivate is actually calling the same SP. If deactivating running into an issue, it's very possible that reactivating records could run into issue too. And if that's the case, it can cause bad data in the DB. That's why I think if we run into the issue, we need to look at it manually instead of having API try to fix it.